cpp/src/arrow/util/value_parsing.h

```diff
@@ -273,6 +273,96 @@ inline bool ParseUnsigned(const char* s, size_t length, uint64_t* out) {
 #undef PARSE_UNSIGNED_ITERATION
 #undef PARSE_UNSIGNED_ITERATION_LAST
 
+#define PARSE_HEX_ITERATION(C_TYPE)                                     \
+  if (length > 0) {                                                     \
+    char val = *s;                                                      \
+    s++;                                                                \
+    result = static_cast<C_TYPE>(result << 4);                          \
+    length--;                                                           \
+    if (val >= '0' && val <= '9'){                                      \
+      result = static_cast<C_TYPE>(result | (val -'0'));                \
+    } else if (val >= 'A' && val <= 'F'){                               \
+      result = static_cast<C_TYPE>(result | (val -'A' + 10));           \
+    } else if (val >= 'a' && val <= 'f'){                               \
+      result = static_cast<C_TYPE>(result | (val -'a' + 10));           \
+    } else {                                                            \
+      /* Non-digit */                                                   \
+      return false;                                                     \
+    }                                                                   \
+  } else {                                                              \
+    break;                                                              \
+  }
+
+
+inline bool ParseHex(const char* s, size_t length, uint8_t* out) {
+  uint8_t result = 0;
```

Comment:
Instead of having multiple function for each type, I suggest to use templates and given that each function is conceptually doing the same operation but different number of times based on the output data type width, use a for-loop instead.
```c++
template <typename T>
inline bool ParseHex(...) {
   ...
   for (int i = 0; i < sizeof(T) * 2; ++i) {
      ParseHexIteration<T>(result);
   }
   ...
}
```

Reply:
As you prefer, but I would be fine with keeping this a macro in this case. Note there is a control flow inside the macro so you cannot turn it into such a trivial loop.

Reply:
I turned it into a loop, it works fine